fix(fs/shutdown): freeze nexus when faulted

tiagolobocastro opened the pull request:
When the last healthy child of a nexus is faulted, errors are propagated up the stack causing the filesystem to shutdown itself down.

Instead I propose that we simply leave the nexus in a paused state, allowing the initiator to start the re-connection cycle, giving us some “time” to recreate the nexus again when the replica is available again.

todo: we might want to add a time-limit after which we resort to failing IOs?

tiagolobocastro committed:
test(nexus/fault): add nexus fault fs shutdown test

Signed-off-by: Tiago Castro <tiagolobocastro@gmail.com>

tiagolobocastro committed:
fix(fs/shutdown): pause nexus when faulted

When the last healthy child of a nexus is faulted, errors are propagated up the stack causing the
filesystem to shutdown itself down.

Instead I propose that we simply leave the nexus in a paused state, allowing the initiator to
start the re-connection cycle, giving us some “time” to recreate the nexus again when the
replica is available again.

todo: we might want to add a time-limit after which we resort to failing IOs?

Signed-off-by: Tiago Castro <tiagolobocastro@gmail.com>

tiagolobocastro committed:
test(nexus/fault): add nexus ana fault fs shutdown test

Similar to the existing test, but attempts to reconnect a new path, ensuring our
failover strategy is still working.

Signed-off-by: Tiago Castro <tiagolobocastro@gmail.com>

mittachaitu commented:
Hi @tiagolobocastro, I have general question about a scenario where FS goes into read only state due to errors returned back at the time of nexus being marked as faulted. Let's say if target accept bunch of IOs from application among them after submitting IOs to replica if there is failure on first IO, we immediately call pause in replica tear down workflow (which will not accept further IOs from client side). If inflight IOs are returned failure to initiator won't there be chance of marking FS into Read-Only?

tiagolobocastro commented:
> Hi @tiagolobocastro, I have general question about a scenario where FS goes into read only state due to errors returned back at the time of nexus being marked as faulted. Let's say if target accept bunch of IOs from application among them after submitting IOs to replica if there is failure on first IO, we immediately call pause in replica tear down workflow (which will not accept further IOs from client side). If inflight IOs are returned failure to initiator won't there be chance of marking FS into Read-Only?

hmm in this case we're leaving the nexus as paused until it is recreated, so we'll not be succeeded nor failing any IO back to the initiator.

tiagolobocastro committed:
fix(nexus/io): set nvmf subsystem as frozen

Multiple children may be retired at "the same time". In case the first retires "wins" and
pauses the subsystem, if we simply leave it paused at the end, the other retires will remain
stuck in the pause.
Instead, introduce a new state of Frozen, which allows other pauses to execute their paused
code and at the end the subsystem remains frozen.

Signed-off-by: Tiago Castro <tiagolobocastro@gmail.com>

tiagolobocastro committed:
test(nexus/freeze): add subsystem freeze test

Signed-off-by: Tiago Castro <tiagolobocastro@gmail.com>

tiagolobocastro commented:
bors merge

tiagolobocastro commented:
bors merge
